game-core/src/main/java/games/strategy/triplea/delegate/battle/casualty/CasualtyOrderOfLosses.java

```diff
@@ -239,29 +237,34 @@ void clearOolCache() {
     }
     sortedWellEnoughUnitsList.addAll(sortedUnitsList);
     // Cache result and all subsets of the result
-    final List<UnitType> unitTypes = new ArrayList<>();
+    final List<AmphibType> unitTypes = new ArrayList<>();
     for (final Unit u : sortedWellEnoughUnitsList) {
-      unitTypes.add(u.getType());
+      unitTypes.add(AmphibType.of(u.getType(), u.getWasAmphibious()));
     }
-    for (final Iterator<UnitType> it = unitTypes.iterator(); it.hasNext(); ) {
+    for (final Iterator<AmphibType> it = unitTypes.iterator(); it.hasNext(); ) {
       oolCache.put(key, new ArrayList<>(unitTypes));
-      final UnitType unitTypeToRemove = it.next();
+      final AmphibType unitTypeToRemove = it.next();
       targetTypes.remove(unitTypeToRemove);
       key = computeOolCacheKey(parameters, targetTypes);
       it.remove();
     }
     return sortedWellEnoughUnitsList;
   }
 
-  static String computeOolCacheKey(final Parameters parameters, final List<UnitType> targetTypes) {
+  @Value(staticConstructor = "of")
+  static class AmphibType {
+    UnitType type;
+    boolean isAmphibious;
+  }
+
+  static String computeOolCacheKey(
+      final Parameters parameters, final List<AmphibType> targetTypes) {
     return parameters.player.getName()
         + "|"
         + parameters.battlesite.getName()
         + "|"
         + parameters.combatModifiers.isDefending()
         + "|"
-        + parameters.combatModifiers.isAmphibious()
```

Comment:
I do suspect the cache will be incorrect here. If we fight a battle with non-amphib units, then later those units are amphib, the cache will have a value when we would have preferred to have had a cache miss.

There were already some problems here, and this code is certainly not fully correct, *very* complicated, *and* performance optimized (a really bad combination). 

Any thoughts @trevan ? I suspect that this is a consideration we are building in that just breaks this cache. We would not want a new cache value for every permutation of possibly amphib unit, that really blows it up in size.

Reply:
It's possibly an option to just allow the cache to be incorrect for amphib units.. though not ideal.

Reply:
The cache key includes whether the unit type is amphibious or not.  A non-amphibious marine would create a different cache key than an amphibious marine.  The `AmphibType` keeps track of amphibious or not.

Reply:
I'll change the cache key to also check if there is a marine bonus.  If a unit type doesn't have a marine bonus, then there is no need to care if the unit is amphibious or not.

Reply:
I'm afraid you'll need to check the cache performance before and after and make sure the size of the cache doesn't blow up too much. Maps where this is an issue are world at war or NWO where you can have very large numbers of units.

Reply:
I ran one round of world of war with hard ai.  I printed out the size of the ool cache right before it was cleared.  In case you don't know, the ool cache is cleared after each step, such as combat, production, non-combat move, etc.

The numbers were roughly the same.  In some cases, this new code had more entries and in other cases it had less.

I think the main reason for that is the old code added the `amphibiousLandAttackers` to the cache key.  This new code doesn't use that and instead uses `getWasAmphibious` but it is roughly the same affect.

Reply:
My recollection is faulty, I though it was kept around for longer. That kinda makes it less useful than I had thought.. I really regret I did not stick longer with the rewrite of this code, much of it was deleted... Ah well, one day maybe.

I was thinking for a test to do something like a 300 vs 300 unit battle. We get a cache entry for each unit in the battle per side, but it's no different really as the amphib entry does not really blow that up. The WaW case is significant around round 10-20 of that map, a save game at that location is best, the army stacks get to be pretty outragous with outragous bship stacks (like 30-40 ships, similar number of carriers etc)...

Reply:
I ran WaW with fast ai until I got to the middle of round 10.  I save it and then ran hard ai for a little under a round with the old code and the new code.  I summed up the maximum number of cache entries before each clearing.  With the old code, there were 66,495 cache entries.  With the new code, there were 69,573 cache entries.  So it does add more cache entries but not significantly.

I ran two large scale battles with WaW that I created using edit mode.  Both battles had 100 armored cars, artillery, elites, infantry, marines, mech. infantry, mot. infantry, and tanks.  One battle had a 50/50 split between amphibious and not.  The other battle had 51/49 split between amphibious and not.  I ran the 50/50 battle first.  WaW is low luck so the outcomes of the battles should be fairly deterministic.  With the old code, there were 3566 cache entries after both battles were finished.  With the new code, there were 2199 cache entries.

Reply:
Thank you for the detailed check